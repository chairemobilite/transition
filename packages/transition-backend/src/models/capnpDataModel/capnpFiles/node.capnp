@0xd68f205b1747bdab;

using Cxx = import "/capnp/c++.capnp";
$Cxx.namespace("node");

struct Node {
  uuid                         @0  :Text;
  id                           @1  :UInt32; # unique for the whole network (integer_id in database)
  internalId                   @2  :Text;
  code                         @3  :Text;
  name                         @4  :Text;
  latitude                     @5  :Int32; # divide by 1000000 to get float
  longitude                    @6  :Int32; # divide by 1000000 to get float
  stationUuid                  @7  :Text;
  color                        @8  :Text;
  description                  @9  :Text;
  data                         @10 :Text;
  isEnabled                    @11 :Int8;
  routingRadiusMeters          @12 :Int16;
  defaultDwellTimeSeconds      @13 :Int16;
  transferableNodesUuids       @14 :List(Text);
  transferableNodesTravelTimes @15 :List(Int16); # seconds
  transferableNodesDistances   @16 :List(Int16); # meters
  isFrozen                     @17 :Int8;
}