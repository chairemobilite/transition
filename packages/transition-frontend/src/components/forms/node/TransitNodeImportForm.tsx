/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';
import { withTranslation, WithTranslation } from 'react-i18next';

import ImporterValidator from 'chaire-lib-common/lib/services/importers/ImporterValidator';
import serviceLocator from 'chaire-lib-common/lib/utils/ServiceLocator';
import FileUploaderHOC, { FileUploaderHOCProps } from 'chaire-lib-frontend/lib/components/input/FileUploaderHOC';
import FileImportForm from '../../parts/FileImportForm';

interface NodeImportFormProps extends FileUploaderHOCProps {
    setImporterSelected: (importerSelected: boolean) => void;
}

const NodesImportForm: React.FunctionComponent<NodeImportFormProps & WithTranslation> = (
    props: NodeImportFormProps & WithTranslation
) => {
    const closeImporter = () => props.setImporterSelected(false);

    const onImported = async () => {
        await serviceLocator.collectionManager
            .get('nodes')
            .loadFromServer(serviceLocator.socketEventManager, serviceLocator.collectionManager);
        serviceLocator.collectionManager.refresh('nodes');
        serviceLocator.eventManager.emit(
            'map.updateLayer',
            'transitNodes',
            serviceLocator.collectionManager.get('nodes').toGeojson()
        );
        serviceLocator.eventManager.emit('transferableNodes.dirty');
        serviceLocator.eventManager.emit('progress', { name: 'Importing', progress: 1.0 });
        closeImporter();
    };

    React.useEffect(() => {
        props.addEventListeners();
        serviceLocator.socketEventManager.on('importer.nodesImported', onImported);
        return () => {
            props.removeEventListeners();
            serviceLocator.socketEventManager.off('importer.nodesImported', onImported);
        };
    }, []);

    return (
        <FileImportForm
            validator={props.validator as ImporterValidator}
            pluralizedObjectsName={'nodes'}
            fileNameWithExtension={'nodes.geojson'}
            fileUploader={props.fileUploader}
            fileImportRef={props.fileImportRef}
            onChange={props.onChange}
            acceptsExtension={'.json,.geojson'}
            label={props.t('main:GeojsonFile')}
            closeImporter={closeImporter}
        />
    );
};

export default FileUploaderHOC(withTranslation(['transit', 'main'])(NodesImportForm), ImporterValidator);
