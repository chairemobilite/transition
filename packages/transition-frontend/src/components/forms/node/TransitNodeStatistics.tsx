/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';
import _uniq from 'lodash.uniq';
import { withTranslation, WithTranslation } from 'react-i18next';

import { _isBlank } from 'chaire-lib-common/lib/utils/LodashExtensions';
import { roundToDecimals } from 'chaire-lib-common/lib/utils/MathUtils';
import Node from 'transition-common/lib/services/nodes/Node';
import Line from 'transition-common/lib/services/line/Line';
import Path from 'transition-common/lib/services/path/Path';
import PathCollection from 'transition-common/lib/services/path/PathCollection';
import serviceLocator from 'chaire-lib-common/lib/utils/ServiceLocator';

const SimpleRow: React.FunctionComponent<{ header: string; value?: string | number; isHeader?: boolean }> = ({
    header,
    value = '',
    isHeader = false
}) => {
    return (
        <tr>
            <th className={isHeader ? '_header' : ''}>{header}</th>
            <td>{value}</td>
        </tr>
    );
};

const SingleColumn: React.FunctionComponent<{ header: string; values?: string[] | number[]; isHeader?: boolean }> = ({
    header,
    values = []
}) => {
    const cellValues = values.map((value, index) => {
        return (
            <tr key={`cell${index}`}>
                <td>{value}</td>
            </tr>
        );
    });

    return (
        <React.Fragment>
            <tr key="cellHeader">
                <th className="_header">{header}</th>
            </tr>
            {cellValues}
        </React.Fragment>
    );
};

interface NodeStatsProps extends WithTranslation {
    node: Node;
    associatedPathIds: string[] | undefined;
    pathCollection: PathCollection;
}

const TransitNodeStatistics: React.FunctionComponent<NodeStatsProps> = (props: NodeStatsProps) => {
    const { associatedPathIds, pathCollection } = props;
    const associatedLineIds: string[] = [];
    const pathValues =
        associatedPathIds?.map((associatedPathId) => {
            const pathGeojson = pathCollection.getById(associatedPathId);
            if (pathGeojson) {
                const path = new Path(pathGeojson.properties, false, serviceLocator.collectionManager);
                const direction = path.getAttributes().direction;
                const line = path.getLine() as Line;
                associatedLineIds.push(line.getId());
                const agency = line.getAgency();
                if (agency && line) {
                    return `${agency.getAttributes().acronym} | ${line.getAttributes().shortname} | ${props.t(
                        `transit:transitPath:directions:${direction}`
                    )} • ${path.getShortenedId()}`;
                } else {
                    return path.getShortenedId();
                }
            } else {
                return associatedPathId;
            }
        }) || [];

    const countUniqLines = _uniq(associatedLineIds).length;

    return (
        <table className="_statistics">
            <tbody>
                <SingleColumn
                    header={`${props.t('transit:transitNode:AssociatedPaths')} • ${associatedPathIds?.length ||
                        0} ${props.t('transit:transitPath:paths')} • ${countUniqLines} ${props.t(
                        'transit:transitLine:lines'
                    )}`}
                    values={pathValues}
                />
            </tbody>
        </table>
    );
};

export default withTranslation(['transit'])(TransitNodeStatistics);
