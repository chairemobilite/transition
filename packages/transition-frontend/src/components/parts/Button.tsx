/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';
import { withTranslation, WithTranslation } from 'react-i18next';

import ButtonCell, { ButtonCellWithConfirm } from './ButtonCell';

interface ButtonProps extends WithTranslation {
    isSelected: boolean;
    children: React.ReactNode;
    key: string;
    /**
     * Whether to flush align to the right the first edit/duplicate/delete
     * action cell. This should be true if no other cell has been flushed to the
     * right. Default: false
     */
    flushActionButtons?: boolean;
    /**
     * If the object can be selected, an edit icon will be shown in the Button.
     * This handler will also be the default handler when the button is clicked
     * if there are no other
     */
    onSelect?: { handler: React.MouseEventHandler; altText?: string };
    /**
     * If the object can be delete, a delete icon will be shown in the Button.
     * It will show a confirmation modal with the message and the delete handler
     * will be called only if the delete is confirmed.
     */
    onDelete?: { handler: React.MouseEventHandler; altText?: string; message: string };
    /**
     * If the object can be duplicated, a duplication icon will be shown in the
     * Button.
     */
    onDuplicate?: { handler: React.MouseEventHandler; altText: string };
}

/**
 * A list element <li> formatted as a Button. Typically to be used in lists of
 * object to be represented with cell-like elements. Children can be added to
 * the props, they will be displayed at the beginning of the button and
 * additional buttons will be added to edit, duplicate and delete the
 * represented object, depending on the props passed.
 * @param props
 * @returns
 */
const Button: React.FunctionComponent<ButtonProps> = (props: ButtonProps) => {
    const deleteText = props.onDelete ? props.onDelete.altText || props.t('main:Delete') : undefined;
    const flushActionButtons = props.flushActionButtons || false;

    return (
        <React.Fragment key={props.key}>
            <li
                className={`_list${props.isSelected === true ? ' _active' : ''}`}
                onClick={props.onSelect ? props.onSelect.handler : undefined}
                key="containerLink"
            >
                {props.children}
                {props.onSelect && (
                    <ButtonCell
                        key={`${props.key}edit`}
                        alignment={flushActionButtons ? 'flush' : 'right'}
                        title={props.onSelect.altText || props.t('main:Edit')}
                    >
                        <img
                            className="_icon-alone"
                            src={'/dist/images/icons/interface/edit_white.svg'}
                            alt={props.onSelect.altText || props.t('main:Edit')}
                        />
                    </ButtonCell>
                )}
                {props.onDuplicate && (
                    <ButtonCell
                        key={`${props.key}duplicate`}
                        alignment={flushActionButtons && !props.onSelect ? 'flush' : 'right'}
                        onClick={props.onDuplicate.handler}
                        title={props.onDuplicate.altText}
                    >
                        <img
                            className="_icon"
                            src={'/dist/images/icons/interface/copy_white.svg'}
                            alt={props.onDuplicate.altText}
                        />
                    </ButtonCell>
                )}
                {props.onDelete && (
                    <ButtonCellWithConfirm
                        key={`${props.key}delete`}
                        alignment={flushActionButtons && !props.onSelect && !props.onDuplicate ? 'flush' : 'right'}
                        onClick={props.onDelete.handler}
                        title={deleteText}
                        confirmButtonColor="red"
                        confirmButtonText={props.onDelete.altText || props.t('main:Delete')}
                        message={props.onDelete.message}
                    >
                        <img
                            className="_icon-alone"
                            src={'/dist/images/icons/interface/delete_white.svg'}
                            alt={deleteText}
                        />
                    </ButtonCellWithConfirm>
                )}
            </li>
            <li className="_clear" key="clearer"></li>
        </React.Fragment>
    );
};

export default withTranslation('main')(Button);
