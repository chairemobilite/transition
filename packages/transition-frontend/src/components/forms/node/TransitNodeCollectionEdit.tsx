/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';
import Collapsible from 'react-collapsible';
import { withTranslation, WithTranslation } from 'react-i18next';
import { faTrashAlt } from '@fortawesome/free-solid-svg-icons/faTrashAlt';
import { faArrowLeft } from '@fortawesome/free-solid-svg-icons/faArrowLeft';
import { faCheckCircle } from '@fortawesome/free-solid-svg-icons/faCheckCircle';
import _cloneDeep from 'lodash/cloneDeep';
import _get from 'lodash/get';
import _toString from 'lodash/toString';
import { featureCollection as turfFeatureCollection } from '@turf/turf';

import InputStringFormatted from 'chaire-lib-frontend/lib/components/input/InputStringFormatted';
import InputColor from 'chaire-lib-frontend/lib/components/input/InputColor';
import InputWrapper from 'chaire-lib-frontend/lib/components/input/InputWrapper';
import { InputCheckboxBoolean } from 'chaire-lib-frontend/lib/components/input/InputCheckbox';
import Button from 'chaire-lib-frontend/lib/components/input/Button';
import { _isBlank, _toInteger } from 'chaire-lib-common/lib/utils/LodashExtensions';
import FormErrors from 'chaire-lib-frontend/lib/components/pageParts/FormErrors';
import serviceLocator from 'chaire-lib-common/lib/utils/ServiceLocator';
import ConfirmModal from 'chaire-lib-frontend/lib/components/modal/ConfirmModal';
import LoadingPage from 'chaire-lib-frontend/lib/components/pages/LoadingPage';
import Node from 'transition-common/lib/services/nodes/Node';
import Preferences from 'chaire-lib-common/lib/config/Preferences';

interface TransitNodeCollectionEditProps extends WithTranslation {
    nodes: Node[];
    onBack: (e: React.MouseEvent) => void;
}

type editableFields = 'color' | 'routing_radius_meters' | 'default_dwell_time_seconds' | 'is_frozen';

interface TransitNodeCollectionEditState {
    nodeErrors?: string[];
    color?: string;
    routing_radius_meters?: number;
    default_dwell_time_seconds?: number;
    is_frozen?: boolean;
    confirmModalDeleteIsOpen: boolean;
    confirmModalBackIsOpen: boolean;
    editableNodeCount: number;
}

// TODO Add simple unit tests for this form, before fixing other todos, it's quite straightforward to do
class TransitNodeCollectionEdit extends React.Component<
    TransitNodeCollectionEditProps,
    TransitNodeCollectionEditState
> {
    constructor(props: TransitNodeCollectionEditProps) {
        super(props);

        // TODO Values should be blank if they are not exactly the same for all nodes, otherwise, they take the common value
        // TODO How to handle frozen nodes in the selection, the user should be warned there are frozen nodes and if the is_frozen is not explicitly changed, those nodes should not be affected
        const node = this.props.nodes[0];
        // Set the nodes as editing
        const unfrozenNodes = this.props.nodes.filter((node) => !node.isFrozen());
        unfrozenNodes.forEach((node) => node.startEditing());
        this.state = {
            confirmModalDeleteIsOpen: false,
            confirmModalBackIsOpen: false,
            color: node ? node.attributes.color : undefined,
            routing_radius_meters: node ? node.attributes.routing_radius_meters : undefined,
            default_dwell_time_seconds: node ? node.attributes.default_dwell_time_seconds : undefined,
            is_frozen: node ? node.attributes.is_frozen || undefined : undefined,
            nodeErrors: undefined,
            editableNodeCount: unfrozenNodes.length
        };
    }

    componentDidMount() {
        serviceLocator.eventManager.emit('map.updateLayers', {
            transitNodes250mRadius: turfFeatureCollection([]),
            transitNodes500mRadius: turfFeatureCollection([]),
            transitNodes750mRadius: turfFeatureCollection([]),
            transitNodes1000mRadius: turfFeatureCollection([]),
            transitNodesRoutingRadius: turfFeatureCollection([])
        });
    }

    componentWillUnmount() {
        serviceLocator.selectedObjectsManager.select('nodeCollectionEdit', null);
    }

    private openDeleteConfirmModal = (e: React.MouseEvent) => {
        e.stopPropagation();
        this.setState({
            confirmModalDeleteIsOpen: true
        });
    };

    private closeDeleteConfirmModal = (e: React.MouseEvent) => {
        e.stopPropagation();
        this.setState({
            confirmModalDeleteIsOpen: false
        });
    };

    private onDelete = (e: React.MouseEvent) => {
        // TODO Do not delete used nodes
        e.stopPropagation();
        serviceLocator.eventManager.emit('map.deleteSelectedNodes');
    };

    private openBackConfirmModal = (e: React.MouseEvent) => {
        e.stopPropagation();
        this.setState({
            confirmModalBackIsOpen: true
        });
    };

    private closeBackConfirmModal = (e: React.MouseEvent) => {
        e.stopPropagation();
        this.setState({
            confirmModalBackIsOpen: false
        });
    };

    private onBack = (e: React.MouseEvent) => {
        e.stopPropagation();

        this.props.nodes.forEach((node) => {
            if (!node.isFrozen() || !node.wasFrozen()) {
                if (node.hasChanged()) {
                    node.cancelEditing();
                }
            }
        });
        this.props.onBack(e);
    };

    private onValueChange = (path: editableFields, value: { value: any; valid: boolean }) => {
        // TODO Converting to any because it gives compilation error otherwise
        this.setState({ [path]: value.value } as any);
        if (value.valid) {
            this.props.nodes.forEach((node) => {
                node.set(path, value.value);
                node.validate();
            });
            // TODO Show each error once, but mention the number of nodes with this error
            this.setState({ nodeErrors: this.props.nodes[0].errors });
            if (['color', 'routing_radius_meters'].includes(path)) {
                this.updateLayers();
            }
        }
    };

    private updateLayers() {
        const geojson = this.props.nodes.map((node) => {
            return _cloneDeep(node.toGeojson());
        });
        serviceLocator.eventManager.emit('map.updateLayers', {
            transitNodesSelected: turfFeatureCollection(geojson)
        });
    }

    private onSave = async (_e: React.MouseEvent) => {
        const nodePromises: Promise<any>[] = [];

        this.props.nodes.forEach((node) => {
            if (node.get('is_frozen', false) === false || !node.wasFrozen()) {
                node.validate();
                if (node.isValid) {
                    if (node.hasChanged()) {
                        nodePromises.push(node.save(serviceLocator.socketEventManager));
                    }
                }
            }
        });

        if (nodePromises.length > 0) {
            serviceLocator.eventManager.emit('progress', { name: 'SavingNode', progress: 0.0 });
            // TODO here to handle rejected promises properly
            await Promise.allSettled(nodePromises);
            serviceLocator.collectionManager.refresh('nodes');
            serviceLocator.eventManager.emit('map.updateLayers', {
                transitNodes: serviceLocator.collectionManager.get('nodes').toGeojson(),
                transitNodesSelected: turfFeatureCollection([]),
                transitNodes250mRadius: turfFeatureCollection([]),
                transitNodes500mRadius: turfFeatureCollection([]),
                transitNodes750mRadius: turfFeatureCollection([]),
                transitNodes1000mRadius: turfFeatureCollection([]),
                transitNodesRoutingRadius: turfFeatureCollection([])
            });
            serviceLocator.eventManager.emit('progress', { name: 'SavingNode', progress: 1.0 });
        }

        if (this.state.nodeErrors && this.state.nodeErrors.length === 0) {
            serviceLocator.eventManager.emit('map.deleteSelectedPolygon');
            serviceLocator.selectedObjectsManager.select('node', null);
        }
    };

    render() {
        if (this.props.nodes.length === 0) {
            return <LoadingPage />;
        }

        const node = this.props.nodes[0];
        const isFrozen = node.getAttributes().is_frozen || false;
        const nodeId = node.getId();
        const hasPaths = node.hasPaths();
        const isContainSelectedFrozenNodes = serviceLocator.selectedObjectsManager.get('isContainSelectedFrozenNodes');
        const changed = this.props.nodes.findIndex((node) => node.hasChanged()) !== -1;

        return (
            <form id="tr__form-transit-node-multi" className="tr__form-transit-node-edit apptr__form">
                <h3>
                    {this.props.t('transit:transitNode:editSelectedNodes', { count: this.state.editableNodeCount })}
                </h3>
                <Collapsible trigger={this.props.t('form:basicFields')} open={true} transitionTime={100}>
                    <div className="tr__form-section">
                        {isFrozen !== true && (
                            <InputWrapper label={this.props.t('transit:transitNode:Color')}>
                                <InputColor
                                    id={`formFieldTransitNodeEditColor${nodeId}`}
                                    value={this.state.color}
                                    defaultColor={_get(Preferences.current, 'transit.nodes.defaultColor', '#0086FF')}
                                    onValueChange={(e) =>
                                        this.onValueChange('color', { value: e.target.value, valid: true })
                                    }
                                />
                            </InputWrapper>
                        )}
                        <InputWrapper
                            label={this.props.t('transit:transitNode:RoutingRadiusMeters')}
                            help={this.props.t('transit:transitNode:NodeRadiusExampleDistance')}
                        >
                            <InputStringFormatted
                                id={`formFieldTransitNodeEditRoutingRadiusMeters${nodeId}`}
                                disabled={isFrozen}
                                stringToValue={_toInteger}
                                valueToString={_toString}
                                value={this.state.routing_radius_meters}
                                onValueUpdated={(value) => this.onValueChange('routing_radius_meters', value)}
                            />
                        </InputWrapper>
                        <InputWrapper label={this.props.t('transit:transitNode:DefaultDwellTimeSeconds')}>
                            <InputStringFormatted
                                id={`formFieldTransitNodeEditDefaultDwellTimeSeconds${nodeId}`}
                                disabled={isFrozen}
                                stringToValue={_toInteger}
                                valueToString={_toString}
                                value={this.state.default_dwell_time_seconds}
                                onValueUpdated={(value) => this.onValueChange('default_dwell_time_seconds', value)}
                            />
                        </InputWrapper>
                    </div>
                </Collapsible>

                <Collapsible trigger={this.props.t('form:advancedFields')} transitionTime={100}>
                    <div className="tr__form-section">
                        <InputWrapper label={this.props.t('main:Locked')}>
                            <InputCheckboxBoolean
                                id={`formFieldTransitNodeEditIsFrozen${nodeId}`}
                                label=" "
                                isChecked={node.isFrozen()}
                                onValueChange={(e) =>
                                    this.onValueChange('is_frozen', { value: e.target.value, valid: true })
                                }
                            />
                        </InputWrapper>
                    </div>
                </Collapsible>
                {isContainSelectedFrozenNodes && (
                    <FormErrors errors={['transit:transitNode:editFrozenSelectedNodesWarning']} errorType="Warning" />
                )}
                {this.state.nodeErrors && <FormErrors errors={this.state.nodeErrors} />}

                <div className="tr__form-buttons-container">
                    <span title={this.props.t('main:Back')}>
                        <Button
                            key="back"
                            color="blue"
                            icon={faArrowLeft}
                            iconClass="_icon-alone"
                            label=""
                            onClick={changed ? this.openBackConfirmModal : this.onBack}
                        />
                    </span>
                    <span title={this.props.t('main:Save')}>
                        <Button icon={faCheckCircle} iconClass="_icon-alone" label="" onClick={this.onSave} />
                    </span>
                    {isFrozen !== true && !hasPaths && (
                        <span title={this.props.t('main:Delete')}>
                            <Button
                                icon={faTrashAlt}
                                iconClass="_icon-alone"
                                label=""
                                color="red"
                                onClick={this.openDeleteConfirmModal}
                            />
                        </span>
                    )}
                </div>
                {this.state.confirmModalDeleteIsOpen && (
                    <ConfirmModal
                        isOpen={true}
                        title={this.props.t('transit:transitNode:ConfirmMultipleDelete')}
                        confirmAction={this.onDelete}
                        confirmButtonColor="red"
                        confirmButtonLabel={this.props.t('transit:transitNode:MultipleDelete')}
                        closeModal={this.closeDeleteConfirmModal}
                    />
                )}
                {this.state.confirmModalBackIsOpen && (
                    <ConfirmModal
                        isOpen={true}
                        title={this.props.t('main:ConfirmBackModal')}
                        confirmAction={this.onBack}
                        confirmButtonColor="blue"
                        confirmButtonLabel={this.props.t('main:DiscardChanges')}
                        cancelButtonLabel={this.props.t('main:Cancel')}
                        closeModal={this.closeBackConfirmModal}
                    />
                )}
            </form>
        );
    }
}

export default withTranslation(['transit', 'main', 'form', 'notifications'])(TransitNodeCollectionEdit);
