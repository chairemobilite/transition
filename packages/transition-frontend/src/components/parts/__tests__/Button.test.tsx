/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import * as React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';

import Button from '../Button';
import { create } from 'node:domain';

// Mock react-markdown and remark-gfm as they use syntax not supported by jest
jest.mock('react-markdown', () => 'Markdown');
jest.mock('remark-gfm', () => 'remark-gfm');

const mockSelect = jest.fn();
const mockDuplicate = jest.fn();
const mockDelete = jest.fn();
const duplicateString = 'duplicate';
const selectString = 'select';
const deleteString = 'delete';

beforeEach(() => {
    mockSelect.mockClear();
    mockDuplicate.mockClear();
    mockDelete.mockClear();
});

test('Minimal props', () => {
    const { container } = render(<Button
        key='key'
        isSelected={true}
    >test</Button>);
    expect(container).toMatchSnapshot();
});

test('All props', () => {
    const { container } = render(<Button
        key='key'
        isSelected={true}
        flushActionButtons={false}
        onSelect={{ handler: mockSelect }}
        onDuplicate={{ handler: mockDuplicate, altText: duplicateString }}
        onDelete={{ handler: mockDelete, message: 'deleting' }}
    >test</Button>);
    expect(container).toMatchSnapshot();
});

test('Flush action buttons', () => {
    const { container } = render(<Button
        key='key'
        isSelected={true}
        flushActionButtons={true}
        onSelect={{ handler: mockSelect }}
        onDuplicate={{ handler: mockDuplicate, altText: duplicateString }}
        onDelete={{ handler: mockDelete, message: 'deleting' }}
    >test</Button>);
    expect(container).toMatchSnapshot();
});

test('Flush action buttons, just delete', () => {
    const { container } = render(<Button
        key='key'
        isSelected={true}
        flushActionButtons={true}
        onDelete={{ handler: mockDelete, message: 'deleting' }}
    >test</Button>);
    expect(container).toMatchSnapshot();
});

test('Not selected, with altText', () => {
    const { container } = render(<Button
        key='key'
        isSelected={false}
        onSelect={{ handler: mockSelect, altText: selectString }}
        onDuplicate={{ handler: mockDuplicate, altText: duplicateString }}
        onDelete={{ handler: mockDelete, altText: deleteString, message: 'deleting' }}
    >test</Button>);
    expect(container).toMatchSnapshot();
});

test('Test select call', () => {
    const { getByText } = render(
        <Button
            key='key'
            isSelected={true}
            onSelect={{ handler: mockSelect }}
        >test</Button>);
    const input = getByText('test') as HTMLInputElement;
    fireEvent.click(input);
    expect(mockSelect).toHaveBeenCalledTimes(1);
});

test('Test duplicate call', () => {
    mockDuplicate.mockImplementationOnce((e) => {
        e.stopPropagation();
    });
    const { getByAltText } = render(
        <Button
            key='key'
            isSelected={true}
            onSelect={{ handler: mockSelect }}
            onDuplicate={{ handler: mockDuplicate, altText: duplicateString }}
        >test</Button>);
    const input = getByAltText(duplicateString) as HTMLInputElement;
    fireEvent.click(input);
    expect(mockDuplicate).toHaveBeenCalledTimes(1);
    expect(mockSelect).not.toHaveBeenCalled();
});

test('Test delete click, cancel', () => {
    mockDuplicate.mockImplementationOnce((e) => {
        e.stopPropagation();
    });
    const { getByAltText, getByText, queryByText } = render(
        <Button
            key='key'
            isSelected={true}
            onDelete={{ handler: mockDelete, altText: deleteString, message: 'deleting' }}
        >test</Button>);
    const input = getByAltText(deleteString) as HTMLInputElement;
    const cancelButton = queryByText('main:Cancel');
    expect(cancelButton).toBeFalsy();
    fireEvent.click(input);

    const cancelButtonAfterClick = getByText('main:Cancel') as HTMLInputElement;
    expect(cancelButtonAfterClick).toBeTruthy();
    fireEvent.click(cancelButtonAfterClick);

    const cancelButtonAfterCancel = queryByText('main:Cancel');
    expect(cancelButtonAfterCancel).toBeFalsy();
    expect(mockDelete).not.toHaveBeenCalled();
});

test('Test delete click, confirm', () => {
    mockDuplicate.mockImplementationOnce((e) => {
        e.stopPropagation();
    });
    const { getByAltText, getByText, queryByText } = render(
        <Button
            key='key'
            isSelected={true}
            onDelete={{ handler: mockDelete, altText: deleteString, message: 'deleting' }}
        >test</Button>);
    const input = getByAltText(deleteString) as HTMLInputElement;
    const confirmButton = queryByText(deleteString);
    expect(confirmButton).toBeFalsy();
    fireEvent.click(input);

    const confirmButtonAfterClick = getByText(deleteString) as HTMLInputElement;
    expect(confirmButtonAfterClick).toBeTruthy();
    fireEvent.click(confirmButtonAfterClick);

    const confirmButtonAfterConfirm = queryByText(deleteString);
    expect(confirmButtonAfterConfirm).toBeFalsy();
    expect(mockDelete).toHaveBeenCalledTimes(1);
});
