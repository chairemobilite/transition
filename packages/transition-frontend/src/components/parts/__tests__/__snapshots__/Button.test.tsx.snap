// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`All props 1`] = `
[
  <li
    className="_list _active"
    onClick={[MockFunction]}
  >
    test
    <span
      className="_list-group _right"
      title="main:Edit"
    >
      <span
        className="_list-element"
      >
        <img
          alt="main:Edit"
          className="_icon-alone"
          src="/dist/images/icons/interface/edit_white.svg"
        />
      </span>
    </span>
    <span
      className="_list-group _right"
      onClick={[MockFunction]}
      title="duplicate"
    >
      <span
        className="_list-element"
      >
        <img
          alt="duplicate"
          className="_icon"
          src="/dist/images/icons/interface/copy_white.svg"
        />
      </span>
    </span>
    <span
      className="_list-group _right"
      onClick={[Function]}
      title="main:Delete"
    >
      <span
        className="_list-element"
      >
        <img
          alt="main:Delete"
          className="_icon-alone"
          src="/dist/images/icons/interface/delete_white.svg"
        />
      </span>
    </span>
  </li>,
  <li
    className="_clear"
  />,
]
`;

exports[`Flush action buttons 1`] = `
[
  <li
    className="_list _active"
    onClick={[MockFunction]}
  >
    test
    <span
      className="_list-group _flush-right _right"
      title="main:Edit"
    >
      <span
        className="_list-element"
      >
        <img
          alt="main:Edit"
          className="_icon-alone"
          src="/dist/images/icons/interface/edit_white.svg"
        />
      </span>
    </span>
    <span
      className="_list-group _right"
      onClick={[MockFunction]}
      title="duplicate"
    >
      <span
        className="_list-element"
      >
        <img
          alt="duplicate"
          className="_icon"
          src="/dist/images/icons/interface/copy_white.svg"
        />
      </span>
    </span>
    <span
      className="_list-group _right"
      onClick={[Function]}
      title="main:Delete"
    >
      <span
        className="_list-element"
      >
        <img
          alt="main:Delete"
          className="_icon-alone"
          src="/dist/images/icons/interface/delete_white.svg"
        />
      </span>
    </span>
  </li>,
  <li
    className="_clear"
  />,
]
`;

exports[`Flush action buttons, just delete 1`] = `
[
  <li
    className="_list _active"
  >
    test
    <span
      className="_list-group _flush-right _right"
      onClick={[Function]}
      title="main:Delete"
    >
      <span
        className="_list-element"
      >
        <img
          alt="main:Delete"
          className="_icon-alone"
          src="/dist/images/icons/interface/delete_white.svg"
        />
      </span>
    </span>
  </li>,
  <li
    className="_clear"
  />,
]
`;

exports[`Minimal props 1`] = `
[
  <li
    className="_list _active"
  >
    test
  </li>,
  <li
    className="_clear"
  />,
]
`;

exports[`Not selected, with altText 1`] = `
[
  <li
    className="_list"
    onClick={[MockFunction]}
  >
    test
    <span
      className="_list-group _right"
      title="select"
    >
      <span
        className="_list-element"
      >
        <img
          alt="select"
          className="_icon-alone"
          src="/dist/images/icons/interface/edit_white.svg"
        />
      </span>
    </span>
    <span
      className="_list-group _right"
      onClick={[MockFunction]}
      title="duplicate"
    >
      <span
        className="_list-element"
      >
        <img
          alt="duplicate"
          className="_icon"
          src="/dist/images/icons/interface/copy_white.svg"
        />
      </span>
    </span>
    <span
      className="_list-group _right"
      onClick={[Function]}
      title="delete"
    >
      <span
        className="_list-element"
      >
        <img
          alt="delete"
          className="_icon-alone"
          src="/dist/images/icons/interface/delete_white.svg"
        />
      </span>
    </span>
  </li>,
  <li
    className="_clear"
  />,
]
`;
